### DISCUSSION TITLE
Where to store a soketi .env file?

### DISCUSSION THREAD

**Discussion by secondmanveran** [soketi/pws#752] [Q&A]
On my local dev server on my Mac soketi is installed at:

```bash
/usr/local/bin/soketi
```

I'm assuming the `.env` should go in `/usr/local/bin` but I've placed it there and soketi is still using the `array` driver when I've specified all the `mysql` credentials and variables in the `.env`

I did of course restart soketi via pm2 and saved the config.

Any idea where I should store the `.env`?

**Comment by nickchomey** [comment]
Perhaps in usr/local/bin/soketi?

**Reply by secondmanveran** [reply]
@nickchomey 

That's not a directory, `/soketi` is the app binary.

Thanks though.

**Reply by nickchomey** [reply]
Ah, sorry. I only just discovered soketi and haven't yet installed it. .env location is one of the things I was wondering about though, as there doesn't seem to be any specific documentation for it. It's just mentioned here https://docs.soketi.app/getting-started/environment-variables

**Reply by nickchomey** [reply]
My only other thought, based on some other env experiences, is to ensure that when you enter `soketi start`, do so from `cd /usr/local/bin`.

Or, perhaps put the env file elsewhere (/usr/xyz/home/.env) and then navigate to `cd /usr/xyz/home` and run `soketi start` from there.

**Reply by hoelderlin2368** [reply]
Can anybody clarify please location and format of env file if you have a working installation. It is not clearly described in the documentation.

Alternatively I am struggling to understand how to set the environment variables correctly on Ubuntu 22.04 here https://docs.soketi.app/getting-started/environment-variables and here https://docs.soketi.app/app-management/sql-drivers/mysql

I am trying to set the MySQL environment variables on my system but I feel it keeps ignoring the settings. Or at least I don't see anything in the logs. Soketi does start successfully and I can reach the websocket ports but how I do I know it reads the environment variables I set?

Can somebody please explain step by step how to make the settings work that are described in the documentation, especially when SQL drivers are required.

See my example I tried to set in the system variables:

/$ env

SHELL=/bin/bash
SOKETI_APP_MANAGER_DRIVER=mysql
SOKETI_DB_MYSQL_PASSWORD=mydbpassword
PWD=/
SOKETI_DB_MYSQL_DATABASE=mydb
LOGNAME=localuser
XDG_SESSION_TYPE=tty
DEFAULT_APP_SECRET=12345
SOKETI_DB_MYSQL_PORT=3306
MOTD_SHOWN=pam
SOKETI_DB_MYSQL_HOST=192.168.100.100
HOME=/home/localuser
LANG=C.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:.tar=01;31:.tgz=01;31:.arc=01;31:.arj=01;31:.taz=01;31:.lha=01;31:.lz4=01;31:.lzh=01;31:.lzma=01;31:.tlz=01;31:.txz=01;31:.tzo=01;31:.t7z=01;31:.zip=01;31:.z=01;31:.dz=01;31:.gz=01;31:.lrz=01;31:.lz=01;31:.lzo=01;31:.xz=01;31:.zst=01;31:.tzst=01;31:.bz2=01;31:.bz=01;31:.tbz=01;31:.tbz2=01;31:.tz=01;31:.deb=01;31:.rpm=01;31:.jar=01;31:.war=01;31:.ear=01;31:.sar=01;31:.rar=01;31:.alz=01;31:.ace=01;31:.zoo=01;31:.cpio=01;31:.7z=01;31:.rz=01;31:.cab=01;31:.wim=01;31:.swm=01;31:.dwm=01;31:.esd=01;31:.jpg=01;35:.jpeg=01;35:.mjpg=01;35:.mjpeg=01;35:.gif=01;35:.bmp=01;35:.pbm=01;35:.pgm=01;35:.ppm=01;35:.tga=01;35:.xbm=01;35:.xpm=01;35:.tif=01;35:.tiff=01;35:.png=01;35:.svg=01;35:.svgz=01;35:.mng=01;35:.pcx=01;35:.mov=01;35:.mpg=01;35:.mpeg=01;35:.m2v=01;35:.mkv=01;35:.webm=01;35:.webp=01;35:.ogm=01;35:.mp4=01;35:.m4v=01;35:.mp4v=01;35:.vob=01;35:.qt=01;35:.nuv=01;35:.wmv=01;35:.asf=01;35:.rm=01;35:.rmvb=01;35:.flc=01;35:.avi=01;35:.fli=01;35:.flv=01;35:.gl=01;35:.dl=01;35:.xcf=01;35:.xwd=01;35:.yuv=01;35:.cgm=01;35:.emf=01;35:.ogv=01;35:.ogx=01;35:.aac=00;36:.au=00;36:.flac=00;36:.m4a=00;36:.mid=00;36:.midi=00;36:.mka=00;36:.mp3=00;36:.mpc=00;36:.ogg=00;36:.ra=00;36:.wav=00;36:.oga=00;36:.opus=00;36:.spx=00;36:.xspf=00;36:
DEFAULT_APP_ID=app-id
DEFAULT_APP_KEY=app-key
SOKETI_DB_MYSQL_USERNAME=mydbuser
SSH_CONNECTION=10.168.165.18 56571 192.168.100.110 22
SOKETI_APP_MANAGER_MYSQL_USE_V2=false
SOKETI_APP_MANAGER_MYSQL_TABLE=apps
LESSCLOSE=/usr/bin/lesspipe %s %s
XDG_SESSION_CLASS=user
TERM=xterm-256color
LESSOPEN=| /usr/bin/lesspipe %s
USER=localadmin
SHLVL=1
XDG_SESSION_ID=1
LC_CTYPE=C.UTF-8
XDG_RUNTIME_DIR=/run/user/1000
SSH_CLIENT=10.168.165.18 56571 22
XDG_DATA_DIRS=/usr/local/share:/usr/share:/var/lib/snapd/desktop
SOKETI_APP_MANAGER_MYSQL_VERSION=8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
SSH_TTY=/dev/pts/0
test=test
OLDPWD=/home
_=/usr/bin/env

**Comment by secondmanveran** [comment]
@nickchomey 

> You may declare soketi server configuration options using environment variables when invoking the soketi server directly on the CLI, or as key-value attributes in an .env file that is placed at the location from where the soketi server command is being run:

It sounds like you may be correct here, but I'm running it with PM2 so I'm wondering it there's a difference in how to load the env.

**Reply by DezyNation** [reply]
Did you find the solution?

**Reply by secondmanveran** [reply]
@DezyNation 

Sorry for the delay, I've been super busy.

I have a production server up and running that serves several development and production websites for myself and my clients. 

The UI is running on a Laravel Jetstream app on Laravel Forge. If you're not familiar, Forge by default creates a user account named forge on the Linux server of your chosen cloud provider, mine is (Ubuntu 22.04.3 LTS (GNU/Linux 5.15.0-84-generic x86_64)) with all the normal privileges you'd expect with a user host account. Additionally it requires an private key to ssh in, root login is disabled, and it does provide sudo access.

All this as a pretense to say, I placed the `.env` file in my user home directory ie: `/home/forge` but I expect that this would be the same on any Linux based distro.

I installed `nvm` to control the node version (currently v18.16.1) as opposed to allowing the server distro to control it, then installed Soketi via `npm` globally. 

I use `PM2` to manage it through the PM2 Keymetrics UI which allows up to 4 servers as I recall. You can read their docs if you're interested in using it. 

Hope this helps.

**Reply by DezyNation** [reply]
Yeah thanks!
Soketi is now working on my VPS as well.
Seems like we have to place the `.env` from where we are running `soketi start`
